### DISCUSSION TITLE
Suggestion On Django Form Submission in Modal.

### DISCUSSION THREAD

**Discussion by ashrafZolkopli** [unpoly/unpoly#317] [Q&A]
Hi all,

Right now I am developing a new application for testing Unpoly 2. I am using Django as my backend server. 
I currently have a List Display view that will show a list of created object. On the page I also have a button that like to a page with a form.
This form is currently being displayed as a modal in the List Display Page. In the form, when ever a validation error occur, The modal is closed and the page is redirected to the create page.

I dont know how or what I should do to keep the page on the modal.

Any idea would be a great help.

sincerely
Ash

**Comment by adam12** [comment]
You need to ensure two things are happening for Unpoly to re-render inside the modal:

1. The form to be [submitted by Unpoly](https://unpoly.com/form-up-submit).
2. The correct status code returned if the form was invalid, which cannot be a 200. Normally a 400 or 422 is appropriate. See linked docs for `up-submit`.

**Comment by ashrafZolkopli** [comment]
Hi adam,

the following is the snippet that I have for the project so far

The button that activate the Modal

![image](https://user-images.githubusercontent.com/59853680/142242522-e520bb49-6aa2-42ee-b2c6-e3f048f6f008.png)


`<a href="{% url 'landing_create' %}" up-layer="new" up-target=".form_class" up-fail-mode="modal"
    up-fail-target=".form_class" class="d-none d-sm-inline-block btn btn-sm btn-primary shadow-sm"><i
        class="fas fa-download fa-sm text-white-50"></i> Create Landing Page</a>`

the form to be rendered in the modal

![image](https://user-images.githubusercontent.com/59853680/142242987-2b7f0b62-9c57-4768-a33c-80e5c19d7fc5.png)


`<form method="post" action="{% url 'landing_create' %}" class='form_class'>`
`
<form method="post" action="{% url 'landing_create' %}" class='form_class'>
    {% csrf_token %}

    {% for field in form %}
    {{ field|attr:"up-validate" |as_crispy_field }}
    {% endfor %}
    <button type="submit" class="btn btn-primary">Create Landing Page</button>
</form>`


also my view for the form submission

`
def LandingCreatePage(request):
    user = request.user
    x_validate = request.headers.get('X-Up-Validate', None)
    print(x_validate)
    form = LandingForm()
    if request.method == "POST":
        form = LandingForm(request.POST or None, user=user)
        context = {
            'user': user,
            "form": form,
        }
        if form.is_valid() and x_validate is None:
            landing_form = form.save(commit=False)
            landing_form.user = user
            landing_form.save()
            return redirect("landing_list")

    context = {
        'user': user,
        "form": form,
    }

    if request.method == "POST":
        return render(request, "Sortable/LandingFormView.html", context, status=400)
    return render(request, "Sortable/LandingFormView.html", context)`

When the form is invalid, it does send a 400 responds to Unpoly, however the page will reload to its own webpage

![image](https://user-images.githubusercontent.com/59853680/142242662-fcde6589-8e64-499e-b90a-c35d76375774.png)

**Reply by ashrafZolkopli** [reply]
Sorry for the wrong formating, I dont know how to add code in the github comment section

**Reply by adam12** [reply]
I don't see an `up-submit` attribute on your `<form>` tag.

**Reply by ashrafZolkopli** [reply]
Sorry for the late reply, Your answer is actually what I needed to make it work. Thanks you so much
